pymeasure/instruments/common_base.py

```diff
@@ -0,0 +1,569 @@
+        for attr in self.fset_params_list:
+            attr_instance_name = self.prefix + "_".join([self.name, attr])
+            if hasattr(obj, attr_instance_name):
+                kwargs[attr] = getattr(obj, attr_instance_name)
+        self.fset(obj, value, **kwargs)
+
+    def __set_name__(self, owner, name):
+        self.name = name
+
+
+class CommonBase:
+    """Base class for instruments and channels.
+
+    This class contains everything needed for pymeasure's property creator
+    :meth:`control` and its derivatives :meth:`measurement` and :meth:`setting`.
+    """
+
+    # Variable holding the list of DynamicProperty parameters that are configurable
+    # by users
+    _fget_params_list = ('get_command',
+                         'values',
+                         'map_values',
+                         'get_process',
+                         'command_process',
+                         'check_get_errors')
+
+    _fset_params_list = ('set_command',
+                         'validator',
+                         'values',
+                         'map_values',
+                         'set_process',
+                         'command_process',
+                         'check_set_errors')
+
+    # Prefix used to store reserved variables
+    __reserved_prefix = "___"
+
+    def __init__(self):
+        self._special_names = self._setup_special_names()
+        # Add children from ChildDescriptors.
+        for item, value in inspect.getmembers(self.__class__):
+            if isinstance(value, ChildDescriptor):
+                if isinstance(value.id, (list, tuple)):
+                    for id, cls in zip(value.id, value.cls):
+                        self.add_child(cls, id, collection=item, **value.kwargs)
+                else:
+                    self.add_child(value.cls, value.id, collection=item, **value.kwargs)
+
+    @staticmethod
+    def children(id, cls, **kwargs):
+        """Describe children of this parent class.
+
+        The children will be added to the parent instance at instantiation with
+        :meth:`add_child`. The variable name (e.g. :code:`channels`) will be
+        used as the `collection` of the children. You may define the attribute
+        prefix via keyword arguments. Normally, do use :code:`channels` as variable
+        and do leave the prefix at the default :code:`"ch"`.
+
+        .. code::
+
+            class SomeInstrument(Instrument):
+                # Three channels of the same type: 'ch_A', 'ch_B', 'ch_C' in 'channels'
+                channels = CommonBase.children(["A", "B", "C"], ChildClass)
+                # Two functions of different types: 'fn_power', 'fn_voltage' in 'functions'
+                functions = CommonBase.children(["power", "voltage"],
+                                                (PowerChannel, VoltageChannel), prefix="fn")
+                # A control without a prefix: 'motor'
+                motor = CommonBase.children(None, MotorControl)
```

Comment:
I'm not clear what channel `id` will ultimately be used here? (there's a lot of indirection). Is it `motor`? 
If yes, we are in trouble if the required string in the command is a number (like `1`), because Python identifiers cannot start with a number.

Reply:
I changed the implementation. Telling `prefix=None` will create a "single channel", e.g. `inst.motors`, such that you can give an id if you wish.

Reply:
Good point! I did not consider that you effectively don't *need* to use the id for filling the command string correctly. I guess this comes down to good examples/docs to show users how to do it.